compiler/passes/resolveIntents.cpp

```diff
@@ -272,8 +323,27 @@ IntentTag concreteIntentForArg(ArgSymbol* arg) {
     // to correctly mark const / not const / maybe const.
     return INTENT_REF;
 
-  else
+  else {
+    if (arg->intent == INTENT_CONST) {
+      // No need to warn if the argument intent was going to be converted to
+      // `const in`
+      if (constIntentForType(arg->type) == INTENT_CONST_REF) {
+        // Want this to be an unstable warning, and not trigger for internal or
+        // standard modules
+        auto mod = fn->getModule();
+        bool shouldWarnInternal = (mod->modTag == MOD_INTERNAL &&
+                                   fWarnUnstableInternal);
+        bool shouldWarnStandard = (mod->modTag == MOD_STANDARD &&
+                                   fWarnUnstableStandard);
+        if (fWarnUnstable && (shouldWarnInternal || shouldWarnStandard ||
+                              mod->modTag == MOD_USER) &&
+            !fNoConstArgChecks) {
+          warnForConstIntent(arg);
+        }
+      }
+    }
     return concreteIntent(arg->intent, arg->type);
+  }
```

Comment:
IMO resolveIntents is a strange place to put this checking but I can see why you did it. I don't think it's a big deal, but resolveIntents does not usually do much AST transformation. I'd have probably attached this checking to `callDestructors` if it were me.

Note that it does not have to be in `resolveIntents` because `ArgSymbol` has a field `originalIntent` that you can use to drive the checking at some later point.

To be clear, I am not expecting you to change this in this PR specifically. But a TODO comment would be nice.

Reply:
Oooooh, I missed the `originalIntent` field, that definitely would make it more principled.  I'll add a TODO and a reminder for myself, thanks!